### PR TITLE
JetBrains: add cancelation support

### PR DESCRIPTION
**Pull request by olafurpg**
Fixes https://github.com/sourcegraph/sourcegraph/issues/56042

Previously, we didn't correctly propagate autocomplete cancelation to the agent. This was problematic because we could end up queueing a lot of autocomplete requests making the experience lag. This PR fixes the problem by sending a `$/cancelRequest` notification to the agent as implemented in https://github.com/sourcegraph/cody/pull/787

## Test plan

While using the plugin, confirm that we're sending a lot of `$/cancelRequest` notifications. Also confirm that autocomplete still works as expected.
```
❯ tail -f build/sourcegraph/cody-agent-trace.json  | grep cancelRe
```
My anecdotal experience is that autocomplete feels somewhat snappier with this change, which makes sense if we're cancelling old requests more aggressively.

### CODE REVIEW

**Review by vdavid — APPROVED** [review]
Nice, this one makes sense! (Reminder: don't merge this before we go through https://github.com/sourcegraph/cody/pull/787/)

File: client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/CodyAutocompleteManager.java [hunk 1]
```diff
@@ -127,12 +126,17 @@ public void triggerAutocomplete(
...
     if (triggerKind.equals(InlineCompletionTriggerKind.INVOKE)
         || autoCompleteDocumentContext.isCompletionTriggerValid()) { // TODO: skip this condition
+      CancellationToken cancellationToken = new CancellationToken();
       Callable<CompletableFuture<Void>> callable =
-          () -> triggerAutocompleteAsync(project, editor, offset, textDocument, triggerKind);
+          () ->
+              triggerAutocompleteAsync(
+                  project, editor, offset, textDocument, triggerKind, cancellationToken);
+      ScheduledFuture<CompletableFuture<Void>> scheduledAutocomplete =
+          this.scheduler.schedule(callable, 20, TimeUnit.MILLISECONDS);
```
Comment (chwarwick) [inline client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/CodyAutocompleteManager.java:135]: Is this logic delaying calling the autocomplete 20ms? The agent (vs code's completion logic) also contains a 25ms delay prior to submitting the network requests [here](https://sourcegraph.com/github.com/sourcegraph/cody/-/blob/vscode/src/completions/getInlineCompletions.ts?L200-204).  I suspect the reason it feels a bit quicker is that with both delays it's submitting fewer requests.  When I experimented with submitting fewer it also felt faster but I got fewer suggestions.

**Comment by vdavid** [comment]
Oh, auto-merge was on. Hmm 😅 Not sure if it'll break?

**Comment by olafurpg** [comment]
It shouldn't break if you build with the wrong commit of Cody, you just get warnings about `$/cancelRequest` notifications that have no handler.

File: client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/CodyAutocompleteManager.java (hunk 1 above)
Comment (olafurpg) [reply client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/CodyAutocompleteManager.java:135]: I was thinking the same, and want to remove this delay in a followup commit. I don't think the delay makes sense with the agent.

Comment (olafurpg) [reply client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/CodyAutocompleteManager.java:135]: Opened an issue to track it https://github.com/sourcegraph/sourcegraph/pull/56119#discussion_r1301739289

**Pull request merged**